playwright/src/main/java/com/microsoft/playwright/impl/UrlMatcher.java

```diff
@@ -68,50 +64,79 @@ static String resolveUrl(URL baseUrl, String spec) {
     }
   }
 
-  UrlMatcher(URL base, String url) {
-    this(url, toPredicate(Pattern.compile(globToRegex(resolveUrl(base, url)))).or(s -> url == null || url.equals(s)));
+  UrlMatcher(URL baseURL, String glob) {
+    this(baseURL, glob, null, null);
   }
 
   UrlMatcher(Pattern pattern) {
-    this(pattern, toPredicate(pattern));
+    this(null, null, pattern, null);
   }
   UrlMatcher(Predicate<String> predicate) {
-    this(predicate, predicate);
+    this(null, null, null, predicate);
   }
 
-  private UrlMatcher(Object rawSource, Predicate<String> predicate) {
-    this.rawSource = rawSource;
+  private UrlMatcher(URL baseURL, String glob, Pattern pattern, Predicate<String> predicate) {
+    this.baseURL = baseURL;
+    this.glob = glob;
+    this.pattern = pattern;
     this.predicate = predicate;
   }
 
   boolean test(String value) {
-    return predicate == null || predicate.test(value);
+    return testImpl(baseURL, pattern, predicate, glob, value);
+  }
+
+  private static boolean testImpl(URL baseURL, Pattern pattern, Predicate<String> predicate, String glob, String value) {
+    if (pattern != null) {
+      return pattern.matcher(value).find();
+    }
+    if (predicate != null) {
+      return predicate.test(value);
+    }
+    if (glob != null) {
+      return Pattern.compile(globToRegex(resolveUrl(baseURL, glob))).matcher(value).find();
+    }
+    return true;
   }
 
   @Override
   public boolean equals(Object o) {
     if (this == o) return true;
     if (o == null || getClass() != o.getClass()) return false;
     UrlMatcher that = (UrlMatcher) o;
-    if (rawSource instanceof Pattern && that.rawSource instanceof Pattern) {
-      Pattern a = (Pattern) rawSource;
-      Pattern b = (Pattern) that.rawSource;
-      return a.pattern().equals(b.pattern()) && a.flags() == b.flags();
+    List<Boolean> matches = new ArrayList<>();
+    if (baseURL != null && that.baseURL != null) {
```

Comment:
I don't think we care if the pattern is a result of a base url + glob or just empty base url and a glob as long as the final glob is the same.